cmd/ci-secret-bootstrap/main.go

```diff
@@ -480,30 +618,54 @@ func main() {
 			logrus.WithError(err).Fatal("Failed to logout.")
 		}
 	})
-	defer logrus.Exit(0)
+
 	ctx := context.TODO()
 	var errs []error
 	// errors returned by constructSecrets will be handled once the rest of the secrets have been uploaded
 	secretsMap, err := constructSecrets(ctx, o.config, bwClient, o.maxConcurrency)
 	if err != nil {
 		errs = append(errs, err)
 	}
-	if o.dryRun {
-		tmpFile, err := ioutil.TempFile("", "ci-secret-bootstrapper")
-		if err != nil {
-			logrus.WithError(err).Fatal("failed to create tempfile")
-		}
-		defer tmpFile.Close()
-		logrus.Infof("Dry-Run enabled, writing secrets to %s", tmpFile.Name())
-		if err := writeSecrets(secretsMap, tmpFile); err != nil {
-			errs = append(errs, fmt.Errorf("failed to write secrets on dry run: %w", err))
+
+	// if validate-bitwarden-items-usage is specified we only want to check if all the
+	// bitwarden items are being used in the given configuration.
+	if o.validateBWItemsUsage {
+		for name, value := range getUnusedBWItems(o.config, bwClient, sets.NewString(o.bwAllowItems.Strings()...)) {
+			logger := logrus.WithField("bw_name", name)
+
+			if value.fields.Len() > 0 {
+				logger = logger.WithField("fields", strings.Join(value.fields.List(), ","))
+			}
+
+			if value.attachments.Len() > 0 {
+				logger = logger.WithField("attachments", strings.Join(value.attachments.List(), ","))
+			}
+
+			if value.hasPassword {
+				logger = logger.WithField("login", "password")
+			}
+
+			logger.Info("Unused item")
 		}
 	} else {
-		if err := updateSecrets(o.secretsGetters, secretsMap, o.force); err != nil {
-			errs = append(errs, fmt.Errorf("failed to update secrets: %w", err))
+		if o.dryRun {
+			tmpFile, err := ioutil.TempFile("", "ci-secret-bootstrapper")
+			if err != nil {
+				logrus.WithError(err).Fatal("failed to create tempfile")
+			}
+			defer tmpFile.Close()
+			logrus.Infof("Dry-Run enabled, writing secrets to %s", tmpFile.Name())
+			if err := writeSecrets(secretsMap, tmpFile); err != nil {
+				errs = append(errs, fmt.Errorf("failed to write secrets on dry run: %w", err))
+			}
+		} else {
+			if err := updateSecrets(o.secretsGetters, secretsMap, o.force); err != nil {
+				errs = append(errs, fmt.Errorf("failed to update secrets: %w", err))
+			}
+			logrus.Info("Updated secrets.")
```

Comment:
@droslean can you elaborate a bit about what you imagine when and by whom this will be used?

IMHO it would make sense to both update and check, so we can do this in the same CI job. And the tool definitely needs to error out if it finds unused items

Reply:
Well, since we already have a periodic job, there is no need to add a new one and use the same tool with different args. So what you are suggesting makes total sense to me. I will update the tool accordingly.